checks/apps/quantumespresso/quantumespresso_check.py

```diff
@@ -1,43 +1,86 @@
-# Copyright 2016-2022 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# Copyright 2016-2023 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
 # ReFrame Project Developers. See the top-level LICENSE file for details.
 #
 # SPDX-License-Identifier: BSD-3-Clause
 
+import pathlib
+import re
+import sys
+
 import reframe as rfm
 import reframe.utility.sanity as sn
 
-
-class QuantumESPRESSOCheck(rfm.RunOnlyRegressionTest):
-    scale = parameter(['small', 'large'])
-    variant = parameter(['maint', 'prod'])
-    modules = ['QuantumESPRESSO']
-    executable = 'pw.x'
-    executable_opts = ['-in', 'ausurf.in', '-pd', '.true.']
-    extra_resources = {
-        'switches': {
-            'num_switches': 1
+sys.path.append(str(pathlib.Path(__file__).parent.parent.parent / 'mixins'))
+from extra_launcher_options import ExtraLauncherOptionsMixin
+from sarus_extra_launcher_options import SarusExtraLauncherOptionsMixin
+from cuda_visible_devices_all import CudaVisibleDevicesAllMixin
+
+qe_tests = {
+    # R:total number of MPI ranks, T:threads per rank, P:walltime
+    'Au-surf': {
+        'zen2': {
+            # zen2 cpu nodes: 2sockets, 8 numa, 16c/numa = 128c (no MT)
+            'energy_reference': -11427.09017218,
+            'performance_reference': [{'R': 32, 'T': 4, 'P': 102.0}]
+        },
+        'zen3-4x-gpu-sm_80': {
+            # A100 nodes: 1socket, 4 numa, 16c/numa = 64c (no MT)
+            'energy_reference': -11427.09017218,
+            'performance_reference': [{'R': 4, 'T': 16, 'P': 32.0}]
         }
     }
-    strict_check = False
-    maintainers = ['LM']
-    tags = {'scs'}
+}
+
+
+class QuantumESPRESSOBase(rfm.RunOnlyRegressionTest):
+    energy_tolerance = 1.0e-6
+    executable = 'pw.x'
+    # TODO: tests should all have pw.in as input file
+    executable_opts = ['-in', 'ausurf.in']
 
     @run_after('init')
-    def set_prog_envs_and_tags(self):
-        if self.current_system.name in ['eiger', 'pilatus']:
-            self.valid_prog_environs = ['cpeIntel']
-        else:
-            self.valid_prog_environs = ['builtin']
+    def set_description(self):
+        self.descr = f'QuantumESPRESSO {self.test_name} check'
 
-        self.tags |= {
-            'maintenance' if self.variant == 'maint' else 'production'
-        }
+    @run_before('run')
+    def set_parallel_resources(self):
+        self.skip_if_no_procinfo()
+        processor_info = self.current_partition.processor
+        self.node_label = processor_info.arch
+        if self.current_partition.devices:
+            # device label, for example 4x-gpu-sm_80
+            dev_label = '-'
+            for e in self.current_partition.devices:
+                dev_label = dev_label + str(e.num_devices) + 'x-' + e.type +
```

Comment:
```suggestion
                dev_label = f"{dev_label}{e.num_devices}x-{e.type}-{e.arch}"
```
where do you set `e.num_devices` ?

Reply:
num_devices is part of `devices` dataset

Reply:
you should use f-string here as suggested